### PR TITLE
Prohibit open generic or address types on Expression.Catch

### PR DESCRIPTION
**Pull request by JonHanna**
Contributes to #8081.

cc: @VSadov

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/CatchBlock.cs [hunk 1]
```diff
@@ -158,7 +157,20 @@ public static CatchBlock MakeCatchBlock(Type type, ParameterExpression variable,
         {
             ContractUtils.RequiresNotNull(type, nameof(type));
             ContractUtils.Requires(variable == null || TypeUtils.AreEquivalent(variable.Type, type), nameof(variable));
-            if (variable != null && variable.IsByRef)
+            if (variable == null)
+            {
+                TypeUtils.ValidateType(type, nameof(type));
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/CatchBlock.cs]: It seems that this PR and the two other related ones for `Convert` and `Goto` always combine `ValidateType` and the `IsRef` and `IsPointer` checks. The same checks seem to be missing in other places where `ValidateType` is used (e.g. `TypeAs`), which makes me wonder whether we could move the checks into `ValidateType` in the first place. Thoughts?

Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/CatchBlock.cs]: Might be a good idea, although need to be sure that IsRef is not going to make problems for byref parameters or returns.

**Comment by VSadov** [comment]
LGTM

File: src/System.Linq.Expressions/src/System/Linq/Expressions/CatchBlock.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/CatchBlock.cs]: At the very least they could go through a single method that does all three calls. Other PRs for the same issue have the same pattern with the same opportunity to reduce redundancy. I'm planning to tie them all together when I've all the places such types can sneak in plugged.

**Pull request merged**